port.js

```diff
@@ -5,233 +5,213 @@ const utqueue = require('ut-queue');
-    this.activeDispatchCount = null;
-    this.activeSendCount = null;
-    this.activeReceiveCount = null;
-    this.msgSent = null;
-    this.msgReceived = null;
-    // codec related metrics
-    this.activeEncodeCount = null;
-    this.activeDecodeCount = null;
-    this.bytesSent = null;
-    this.bytesReceived = null;
-    // state properties
-    this.isReady = false;
-    this.isConnected = new Promise(resolve => {
-        this.resolveConnected = resolve;
-    });
-    this.state = 'stopped';
-}
-
-util.inherits(Port, EventEmitter);
-
-Port.prototype.timing = timing;
-Port.prototype.merge = merge;
-
-Port.prototype.init = function init() {
-    this.logFactory && (this.log = this.logFactory.createLog(this.config.logLevel, {name: this.config.id, context: this.config.type + ' port'}, this.config.log));
-
-    if (this.config.metrics !== false && this.bus && this.bus.config.implementation && this.bus.performance) {
-        let measurementName = this.config.metrics || this.config.id;
-        let tags = {
-            host: os.hostname(),
-            impl: this.bus.performance.config.id || this.bus.config.implementation,
-            pid: process.pid
+module.exports = class Port extends EventEmitter {
+    constructor({ utLog, utBus, utError, config } = {}) {
+        super();
+        this.log = {};
+        this.utLog = utLog;
+        this.bus = utBus;
+        this.errors = createErrors(utError);
+        this.config = this.traverse(obj => obj.hasOwnProperty('defaults') && (obj.defaults instanceof Function ? obj.defaults.apply(this) : obj.defaults), config);
+        this.methods = this.traverse(obj => obj.hasOwnProperty('handlers') && (obj.handlers instanceof Function ? obj.handlers.apply(this) : obj.handlers), {});
+        this.sendQueues = utqueue.queues();
+        this.receiveQueues = utqueue.queues();
+        this.counter = null;
+        this.streams = [];
+        this.imported = new Set();
+        // performance metrics
+        this.methodLatency = null;
+        this.portLatency = null;
+        this.activeExecCount = null;
+        this.activeDispatchCount = null;
+        this.activeSendCount = null;
+        this.activeReceiveCount = null;
+        this.msgSent = null;
+        this.msgReceived = null;
+        // codec related metrics
+        this.activeEncodeCount = null;
+        this.activeDecodeCount = null;
+        this.bytesSent = null;
+        this.bytesReceived = null;
+        // state properties
+        this.isReady = false;
+        this.isConnected = new Promise(resolve => {
+            this.resolveConnected = resolve;
+        });
+        this.state = 'stopped';
+    }
+    traverse(prop, initial) {
+        let config = [initial];
+        for (let current = Object.getPrototypeOf(this); current; current = Object.getPrototypeOf(current)) {
+            let value = prop(current);
+            if (value) config.push(value);
+        }
+        return merge(...config.reverse());
```

Comment:
The items in the config array are objects returned by some functions (e.g. by Class.prototype.defaults() or by Class.prototype.handlers()). But 'defaults' or 'handlers' can also be properties instead functions (somebody might decide to do it this way on implementation level as the code will still work). 
In that case this will cause a difficult to spot mutation. 

Maybe better ```return merge({}, ...config.reverse());```

Reply:
Never mind. As it is reversed, the first argument will always be the Port class so we know it is an object returned by a function call.
the order will always be ```Parent <- Child <- external config```, right?